File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/device/sound/SoundOutputSamplesTarget.java
/**
 *
 * @file
 *
 * @brief Playback implementation of SamplesTarget based on AudioTrack functionality
 *
 * @author vitamin.caig@gmail.com
 *
 */

package app.zxtune.device.sound;

import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;
import android.support.annotation.NonNull;

import app.zxtune.Log;
import app.zxtune.sound.SamplesSource.Channels;
import app.zxtune.sound.SamplesSource.Sample;
import app.zxtune.sound.SamplesTarget;

public final class SoundOutputSamplesTarget implements SamplesTarget {

  private static final String TAG = SoundOutputSamplesTarget.class.getName();
  
  private static final int CHANNEL_OUT = AudioFormat.CHANNEL_OUT_STEREO;
  private static final int ENCODING = AudioFormat.ENCODING_PCM_16BIT;
  private static final int DEFAULT_LATENCY = 160;// minimal is ~55
  public static final int STREAM = AudioManager.STREAM_MUSIC;

  private final int bufferSize;
  private AudioTrack target;
  
  private SoundOutputSamplesTarget(int bufferSize, AudioTrack target) {
    this.bufferSize = bufferSize;
    this.target = target;
  }
  
  public static SamplesTarget create() {
    //xmp plugin limits max frequency
    final int freqRate = Math.min(AudioTrack.getNativeOutputSampleRate(STREAM), 48000);
    final int minBufSize = AudioTrack.getMinBufferSize(freqRate, CHANNEL_OUT, ENCODING);
    final int prefBufSize = Channels.COUNT * Sample.BYTES * (DEFAULT_LATENCY * freqRate / 1000);
    final int bufSize = Math.max(minBufSize, prefBufSize);
    Log.d(TAG, "Preparing playback. Freq=%d MinBuffer=%d PrefBuffer=%d BufferSize=%d", freqRate,
        minBufSize, prefBufSize, bufSize);
    final AudioTrack target =
        new AudioTrack(STREAM, freqRate, CHANNEL_OUT, ENCODING, bufSize, AudioTrack.MODE_STREAM);
    return new SoundOutputSamplesTarget(bufSize, target);
  }
  
  @Override
  public int getSampleRate() {
    return target.getSampleRate();
  }
  
  @Override
  public int getPreferableBufferSize() {
    return bufferSize / Sample.BYTES;
  }
  
  @Override
  public synchronized void start() {
    target.play();
  }

  @Override
  public synchronized void writeSamples(@NonNull short[] buffer) throws Exception {
    for (int pos = 0, toWrite = buffer.length; toWrite != 0;) {
      final int written = target.write(buffer, pos, toWrite);
      if (written > 0) {
        pos += written;
        toWrite -= written;
      } else {
        final int state = target.getPlayState();
        if (state == AudioTrack.PLAYSTATE_STOPPED) {
          break;//drain
        } else if (written < 0) {
          throw new Exception("Failed to write samples: " + written);
        }
      }
    }
  }

  @Override
  public synchronized void stop() {
    target.stop();
  }

  @Override
  public synchronized void release() {
    target.release();
    target = null;
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/device/sound/SoundOutputSamplesTarget.kt
/**
 *
 * @file
 *
 * @brief Playback implementation of SamplesTarget based on AudioTrack functionality
 *
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.device.sound

import android.content.Context
import android.content.Intent
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioTrack
import android.media.audiofx.AudioEffect
import app.zxtune.Logger
import app.zxtune.sound.SamplesSource
import app.zxtune.sound.SamplesSource.Sample
import app.zxtune.sound.SamplesTarget

class SoundOutputSamplesTarget private constructor(
    private val bufferSize: Int,
    private val target: AudioTrack,
    private val effectControl: AudioEffectControl,
) : SamplesTarget {

    override val sampleRate: Int
        get() = target.sampleRate
    override val preferableBufferSize: Int
        get() = bufferSize / Sample.BYTES

    @Synchronized
    override fun start() = target.play()

    @Synchronized
    override fun writeSamples(buffer: ShortArray) {
        var pos = 0
        var toWrite = buffer.size
        while (toWrite != 0) {
            val written = target.write(buffer, pos, toWrite)
            if (written > 0) {
                pos += written
                toWrite -= written
            } else {
                when {
                    target.playState == AudioTrack.PLAYSTATE_STOPPED -> break //drain
                    written < 0 -> throw Exception("Failed to write samples: $written")
                }
            }
        }
    }

    @Synchronized
    override fun stop() = target.stop()

    @Synchronized
    override fun release() {
        effectControl.release()
        target.release()
    }

    companion object {
        private val LOG = Logger(SoundOutputSamplesTarget::class.java.name)
        private const val CHANNEL_OUT = AudioFormat.CHANNEL_OUT_STEREO
        private const val ENCODING = AudioFormat.ENCODING_PCM_16BIT
        private const val DEFAULT_LATENCY = 160 // minimal is ~55
        const val STREAM = AudioManager.STREAM_MUSIC

        @JvmStatic
        fun create(ctx: Context): SamplesTarget {
            //xmp plugin limits max frequency
            val freqRate = AudioTrack.getNativeOutputSampleRate(STREAM).coerceAtMost(48000)
            val minBufSize = AudioTrack.getMinBufferSize(freqRate, CHANNEL_OUT, ENCODING)
            val prefBufSize =
                SamplesSource.Channels.COUNT * Sample.BYTES * (DEFAULT_LATENCY * freqRate / 1000)
            val bufSize = maxOf(minBufSize, prefBufSize)
            LOG.d {
                "Preparing playback. Freq=$freqRate MinBuffer=$minBufSize PrefBuffer=$prefBufSize BufferSize=$bufSize"
            }
            val target =
                AudioTrack(STREAM, freqRate, CHANNEL_OUT, ENCODING, bufSize, AudioTrack.MODE_STREAM)
            val effectControl = AudioEffectControl(ctx, target.audioSessionId)
            return SoundOutputSamplesTarget(bufSize, target, effectControl)
        }
    }
}

private class AudioEffectControl(private val ctx: Context, private val sessionId: Int) {

    init {
        send(AudioEffect.ACTION_OPEN_AUDIO_EFFECT_CONTROL_SESSION)
    }

    fun release() = send(AudioEffect.ACTION_CLOSE_AUDIO_EFFECT_CONTROL_SESSION)

    private fun send(action: String) {
        val intent = Intent(action).apply {
            putExtra(AudioEffect.EXTRA_AUDIO_SESSION, sessionId)
            putExtra(AudioEffect.EXTRA_PACKAGE_NAME, ctx.packageName)
        }
        ctx.sendBroadcast(intent)
    }
}
